Switch SQLUserData model to be sparsely populated

esoergel opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
Tweaks the initialization and save methods of `UserData` to only create `SQLUserData` row in the DB if they will be non-empty.  This should have no user-facing impact.  Once user data _has_ been saved to the DB it won't be deleted (until the user is deleted), even if the user data is later cleared, but the `UserData` class should behave the same in either instance.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
The automated tests are pretty thorough.

### Automated test coverage

All public methods are tested

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Save only nontrivial user data

esoergel committed:
Sprinkle a little db integrity in there

Jtang-1 commented on corehq/apps/users/user_data.py:
```diff
@@ -31,27 +31,44 @@ def lazy_init(cls, couch_user, domain):
+                        .get(user_id=self._couch_user.user_id, domain=self.domain))
+        except SQLUserData.DoesNotExist:
+            if self._local_to_user or self.profile_id:
+                SQLUserData.objects.create(
```
should this be assigned to `sql_data`?

esoergel replied:
nope - the `else` block is only executed if the `try` block succeeds, so after this, the method returns.  Another way to write this is:

```python
    @transaction.atomic
    def save(self):
        try:
            existing = (SQLUserData.objects
                        .select_for_update()
                        .get(user_id=self._couch_user.user_id, domain=self.domain))
        except SQLUserData.DoesNotExist:
            existing = None

        if existing:
            existing.data = self._local_to_user
            existing.profile_id = self.profile_id
            existing.save()
        elif self._local_to_user or self.profile_id:
            # Only save to DB if it's non-empty
            SQLUserData.objects.create(
                user_id=self._couch_user.user_id,
                domain=self.domain,
                data=self._local_to_user,
                django_user=self._couch_user.get_django_user(),
                profile_id=self.profile_id,
            )
```

(That's how I originally wrote it, but then switched it to the current version as I thought it looked better)

MartinRiese commented on corehq/apps/users/user_data.py:
```diff
@@ -31,27 +31,44 @@ def lazy_init(cls, couch_user, domain):
-            },
-        )
+        try:
+            sql_data = (SQLUserData.objects
```
what are the () on the outside for?

esoergel replied:
So it can span multiple lines - you'd get a syntax error without them.

MartinRiese commented on corehq/apps/users/user_data.py:
```diff
@@ -31,27 +31,44 @@ def lazy_init(cls, couch_user, domain):
+                    django_user=self._couch_user.get_django_user(),
+                    profile_id=self.profile_id,
+                )
+            # Don't save to DB if there's nothing to save
```
this comment is for the else block if I am not mistaken.

Nit. indent it the same depth

esoergel replied:
it's not for the `else` block on the line below this, it's for the condition where the `except` is triggered, but the `if` above this does not pass.  I can rephrase it and put it on the `if`

esoergel committed:
move and rephrase comment

esoergel commented:
rebased off master to fix the test failure previously on master

esoergel commented:
@Jtang-1 @MartinRiese Any remaining feedback?

MartinRiese reviewed (approved):
Looks good

esoergel committed:
There are django users without reach couch users

MartinRiese reviewed (approved):
Technically it is a two-off migration now ;)

esoergel committed:
Rewrite migration to be faster

This fetches user docs in bulk, doesn't wrap them (which is actually
ideal, as we're trying to extract the raw stored data, not the wrapped
data), and avoids re-fetching the django users when creating SQL data
